### DISCUSSION TITLE
Element lost outside of "if" sentence

### DISCUSSION THREAD

**Discussion by leonardorame** [nlohmann/json#3276] [Q&A] [answered]
Hi, I would like to know why an instance I `push_back()` into an array, created inside an `if` sentence, disappears from the array when I want to access from outside that `if`.

Here's an example:

```cpp
json book;
book["pages"] = json::array();
int max_rows = 10;
json page;
for(int i=0; i<10; i++) {
    if(max_rows == 10) {
        page = json();
        page["paragraphs"] = json::array();
        book["pages"].push_back(page);
        max_rows = 10;
    };
    json paragraph;
    paragraph["word1"] = "one";
    paragraph["word2"] = "two";
    paragraph["word3"] = "three";
    page["paragraphs"].push_back(paragraph);
    max_rows--;
};

std::cout << book.dump() << std::endl;
```
Each page in the array `book["pages"]` should contain the array `paragraphs`, but the dump shows `book["pages"]` as 10 `null` elements.

It looks like the `page = json();` instance is destroyed at the end of the `if` sentence (because it's out of scope). How can I prevent this?.

**Comment by nlohmann** [comment]
There seems to be a logic error. In every loop, you re-initialize `page` right in the `if`.

**Comment by leonardorame** [comment, marked as answer]
Yes, I can re-formulate the code, by using a different variable, but the result is similar:

```cpp
json book;
book["pages"] = json::array();
int max_rows = 10;
json currentpage;
for(int i=0; i<10; i++) {
    if(max_rows == 10) {
        page = json();
        page["paragraphs"] = json::array();
        book["pages"].push_back(page);
        max_rows = 10;
        curentpage = page;
    };
    json paragraph;
    paragraph["word1"] = "one";
    paragraph["word2"] = "two";
    paragraph["word3"] = "three";
    currentpage["paragraphs"].push_back(paragraph);
    max_rows--;
};

std::cout << book.dump() << std::endl;
```

**Reply by leonardorame** [reply]
What I would like to do is accessiong the last element of `book["pages"]`, then add the paragraph to that element.

Something like this:

```
json last_page = book["pages"].back();
last_page["paragraph"].push_back(paragraph);
```

But of course, `back()` doesn't return the last element of the array.

**Reply by nlohmann** [reply]
It does, but it returns a reference, see https://json.nlohmann.me/api/basic_json/back/. You make a copy. Change this to

```cpp
json& last_page = book["pages"].back();
last_page["paragraph"].push_back(paragraph);
```

**Reply by leonardorame** [reply]
Thanks!, I'll try that.

**Reply by leonardorame** [reply]
Thank you very much @nlohmann that worked as expected.
